src/markdown-extensions/extracted-code-blocks/FilterBase.php

```diff
@@ -208,16 +241,134 @@ public function filter(
+    // expensive.
+    if (!Regex\matches(
+      $code,
+      re"/^(namespace|use|const|(async |)function|[a-z ]*class|interface|trait|enum) /m",
+    )) {
+      $code = Str\trim_right($code)
+        |> Str\split($$, "\n")
+        |> Vec\map($$, $line ==> Str\trim_right('  '.$line))
+        |> Str\join($$, "\n");
+      $code = "<<__EntryPoint>>\nfunction _main(): void {\n$code\n}\n";
+    }
+
+    // Insert \init_docs_autoloader() call at the beginning of the
+    // <<__EntryPoint>> function (if there is one).
+    if (
+      !Str\ends_with($path, '.'.self::TYPE_ERRORS) &&
+      Str\contains($code, '<<__EntryPoint>>') &&
+      !Str\contains($code, 'init_docs_autoloader()')
+    ) {
+      $entrypoint_opening_brace = Str\search(
+        $code,
+        '{',
+        Str\search($code, '<<__EntryPoint>>') as nonnull
+      );
+      invariant(
+        $entrypoint_opening_brace is nonnull,
+        'Failed to locate <<__EntryPoint>> opening brace.',
+      );
+      $code = Str\slice($code, 0, $entrypoint_opening_brace + 1).
+        "\n  \\init_docs_autoloader();\n".
+        Str\slice($code, $entrypoint_opening_brace + 1);
+    }
+
+    // Insert `use` clauses for common HSL namespaces. There might be some false
+    // positives here, but that's fine.
+    $uses = vec[];
+    foreach (self::COMMON_HSL_NAMESPACES as $ns) {
+      if (
+        Str\contains($code, $ns.'\\') &&
+        !\preg_match('/^use namespace HH\\\\Lib\\\\.*'.$ns.'/m', $code)
+      ) {
+        $uses[] = $ns;
+      }
+    }
+    if (!C\is_empty($uses)) {
+      $uses_str = Str\join($uses, ', ');
+      if (C\count($uses) > 1) {
+        $uses_str = '{'.$uses_str.'}';
+      }
+      $code = 'use namespace HH\\Lib\\'.$uses_str.";\n\n$code";
+    }
+
+    // Generate a unique namespace name (unless one is already specified),
+    // e.g. HHVM\UserDocumentation\Guides\Hack\GettingStarted\MyFirstProgram.
+    if (
+      !Str\starts_with($code, 'namespace ') &&
+      !Str\contains($code, "\nnamespace ")
+    ) {
+      // Cut off anything after the first period. This makes it possible to
+      // force multiple examples to use the same namespace, e.g. example.hack
+      // and example.inc.hack.
+      $ns = Str\slice($path, 0, Str\search($path, '.') as nonnull)
+        |> Str\split($$, '/')
+        |> Vec\drop($$, C\find_key($$, $v ==> $v === 'guides') as nonnull)
+        |> Vec\map(
+          $$,
+          $part ==> Str\trim_left($part, '0123456789-_')
+            |> Str\strip_suffix($$, '-examples')
+            |> Str\capitalize($$)
+            |> Regex\replace_with( // convert to CamelCase
+              $$,
+              re"/[^a-zA-Z0-9]+([a-zA-Z0-9])/",
+              $match ==> Str\uppercase($match[1]),
+            ),
+        )
+        |> Str\join($$, '\\');
+      $code = 'namespace HHVM\\UserDocumentation\\'.$ns.";\n\n$code";
+    }
+
+    /* TODO: Uncomment later.
```

Comment:
Is this finished?

Reply:
Oh, sorry, I forgot to mention this; I'm planning to uncomment this right after this lands, I just didn't want to do it in the same PR since it would make it even harder to review (my test plan relies on comparing old example files with the newly extracted ones, which would be much harder if these extra comments showed up everywhere)...